pkg/backend/snapshot.go

```diff
@@ -415,26 +465,61 @@ func (sm *SnapshotManager) snap() *deploy.Snapshot {
 	return deploy.NewSnapshot(manifest, resources, operations)
 }
 
+// saveSnapshot persists the current snapshot and optionally verifies it afterwards.
+func (sm *SnapshotManager) saveSnapshot() error {
+	snap := sm.snap()
+	if err := sm.persister.Save(snap); err != nil {
+		return errors.Wrap(err, "failed to save snapshot")
+	}
+	if sm.doVerify {
+		if err := snap.VerifyIntegrity(); err != nil {
+			return errors.Wrapf(err, "failed to verify snapshot")
+		}
+	}
+	return nil
+}
+
 // NewSnapshotManager creates a new SnapshotManager for the given stack name, using the given persister
 // and base snapshot.
 //
 // It is *very important* that the baseSnap pointer refers to the same Snapshot
 // given to the engine! The engine will mutate this object and correctness of the
 // SnapshotManager depends on being able to observe this mutation. (This is not ideal...)
 func NewSnapshotManager(persister SnapshotPersister, baseSnap *deploy.Snapshot) *SnapshotManager {
+	mutationRequests, done := make(chan mutationRequest), make(chan error)
+
 	manager := &SnapshotManager{
 		persister:        persister,
 		baseSnapshot:     baseSnap,
 		dones:            make(map[*resource.State]bool),
 		completeOps:      make(map[*resource.State]bool),
 		doVerify:         true,
-		mutationRequests: make(chan func()),
+		mutationRequests: mutationRequests,
+		done:             done,
 	}
 
+	// True if we have elided writes since the last actual write.
```

Comment:
Seems like this can be declared inside the closure? Doesn't look like it's used outside of it.